articles/200_coexistence_and_migration.md

```diff
@@ -0,0 +1,249 @@
+
+## Goals
+
+### Separately usable
+
+* One version should not be affected by other versions choices / drawbacks.
+
+* ROS 1 users shouldn't be worried about ROS 2 development potentially negatively affecting ROS 1.
+
+  * No changes to existing ROS 1 packages should be necessary to continue to work.
+
+* ROS 2 users shouldn't need to make compromises due to ROS 1 choices / limitations / design decisions.
+
+### Support dual-homed applications
+
+To support creating dual-homed applications like the [ros1_bridge](https://github.com/ros2/ros1_bridge) ROS 1 and ROS 2 must be usable within a single application.
+
+### Minimize migration effort
+
+To ease the migration path the goal is to minimize the effort necessary where possible and when this goal doesn't conflict with the previously mentioned goals.
+
+## Technical Constraints
+
+### Namespace of ROS packages
+
+The first and very important question is about the namespace of ROS packages.
+Should ROS 1 and ROS 2 packages use separate namespaces or should they share the same namespace?
+For example can there be two packages names `foo` - one in ROS 1 and the other in ROS2 or do they have to be named differently.
+
+In the case that they use separate namespaces that does not prevent the release of a package `bar` into both, ROS 1 as well as ROS 2.
+
+Having them share a single namespace makes it difficult for users to distinguish them.
+
+Using separate namespaces allows e.g. to release a package `foo` in ROS 2 which provides the same API (but different implementation) as the ROS 1 package with the same name.
+This way downstream packages don't need to change when switching between the ROS 1 and ROS 2 version of the package.
+
+**Decision:** Use separate namespaces for packages in ROS 1 and ROS 2 (this is currently already the case).
+
+### Namespace of rosdep keys
+
+The next question is very similar and is about the namespace of the `rosdep` keys.
+Should ROS 1 and ROS 2 use a different set of `rosdep` rules or should they share the same database?
+
+#### Same key vs. different key
+
+If a ROS package declares a dependency on the `rosdep` key `python-foo` that maps on e.g. Ubuntu Xenial to the Python 2 package `foo`.
+In order to work for ROS 2 it needs to depend on the Python 3 version of the package instead.
+
+Currently the mapping in `rosdep` only depends on the platform and doesn't consider which ROS distribution is being used.
+
+If the migration to ROS 2 requires the key to be changed to `python3-foo` that implies that any package which depends on Python packages does need a modification in the manifest.
+
+The same conceptional problems already exist in ROS 1.
+When being built on platforms which use Python 3 they already map the keys to Python 3 packages.
+If this problem would be address within `rosdep` to consider the Python version as part of the platform that would allow Python 2 as well as Python 3 packages to be resolved using the same `rosdep` key.
```

Comment:
Currently in rosdep you can specify that you want the python3 or the python2 version using the same key (defining an environment variable on your system). We don't have many rosdep keys providing the python3 version, mostly because it's cumbersome to define them. We discussed with @tfoote recently ways to make this more convenient (https://github.com/ros-infrastructure/rosdep/issues/547).
One of the suggested ideas is to allow rosdep rules like:
```
python-foo:
  ubuntu: [python-foo]
  ubuntu_python3: [python3-foo]
```

the current system forces you to define:
```
python-foo:
  ubuntu:
    trusty: [python-foo]
    trusty_python3: [python3-foo]
    xenial: [python-foo]
    xenial_python3: [python3-foo]
    zesty: [python-foo]
    zesty_python3: [python3-foo]
    ...
```

This way there is no need to change the package.xml because they use the same key `python-foo`
While currently for ros2 we need another key for each package
```
python3-foo:
  ubuntu: [python3-foo]
```
> If this problem would be address within `rosdep` to consider the Python version as part of the platform

Maybe you can clarify @dirk-thomas what you mean by this exactly.
 do you mean "`arch` should always map to python3 and `ubuntu` should always map to python2"
or "allow user to define on their platform that they want to map to the python3 version of the packages and not the python2 one"

> Or would I still be able to specify explicitly if I want my ROS 1 package to depend on the python2 or python3 version of a dependency? 

My reading of this is that you won't be able to specify this in the `package.xml` and that will be configured globally as a trade-of between not changing your package.xml and python3 users able to use python 3 dependencies for their for packages.
I think it's a valid point being raised. How does that work for people using dual-home systems ? if they "have to" define that they want to use python3 for their platform (because using ROS2), will they be able to use ROS1 packages that are compatible only with the python2 version of their dependencies?

If that's the case, how can we mitigate it ?
- by inciting people to either be compatible with both python2 and python3 ?
- by asking them to create python2 only keys and use those if they dont want to upgrade their package?
- something else?

Reply:
The question is "how could rosdep know to resolve Python 2 packages in ROS 1 and Python 3 packages in ROS 2 for the same key"? If the user provides a custom configuration like `xenial_python3` that is already possible but for the ROS 2 case it is desired that it is not necessary to do so.

The Debian packages created for a single ROS distribution have to agree on a Python version. Currently for ROS 1 that is Python 2. But it should be possible to create binary package like `ros-lunar3-*` which choose Python 3 instead. Just mixing isn't supported. Nothing prevents a user to bypass the default Python version and embed whatever is desired in e.g. a shebang line (as long as the line isn't rewritten by setuptools). Currently that is being achieved by having a `*_python3` mapping for some rosdep keys and the user choosing this explicitly.

For "dual-homed" application this is not relevant imo. The user will first install the ROS 1 dependencies in a ROS 1-only environment, then does the same in a ROS 2-only environment. For afterwards running the dual-homed application it won't interact with rosdep anymore.